lib/profiling/nvidia/nvtx_prof.f90

```diff
@@ -0,0 +1,181 @@
+  integer :: col(NUM_COLOURS) = [ Z'0000ff00', Z'000000ff', Z'00ffff00', &
+                                  Z'00ff00ff', Z'0000ffff', Z'00ff0000', &
+                                  Z'00ffffff']
+
+  !> Holds the name of the region being created
+  character(len=256), target :: tempName
+
+  !> Struct passed into nvtxRangePushEx
+  type, bind(C):: nvtxEventAttributes
+     integer(C_INT16_T):: version=1
+     integer(C_INT16_T):: size=48
+     integer(C_INT):: category=0
+     integer(C_INT):: colorType=1   !> NVTX_COLOR_ARGB = 1
+     integer(C_INT):: color
+     integer(C_INT):: payloadType=0 !> NVTX_PAYLOAD_UNKNOWN = 0
+     integer(C_INT):: reserved0
+     integer(C_INT64_T):: payload   !> union uint,int,double
+     integer(C_INT):: messageType=1 !> NVTX_MESSAGE_TYPE_ASCII     = 1 
+     type(C_PTR):: message          !> ASCII char
+  end type nvtxEventAttributes
+
+  interface nvtxRangePush
+     ! Push range with custom label and standard colour
+     subroutine nvtxRangePushA(name) bind(C, name='nvtxRangePushA')
+       use iso_c_binding
+       ! Have to use length of 1 because routine is BIND(C)
+       character(kind=C_CHAR, len=1) :: name
+     end subroutine nvtxRangePushA
+
+     ! Push range with custom label and custom colour
+     subroutine nvtxRangePushEx(event) bind(C, name='nvtxRangePushEx')
+       use iso_c_binding
+       import :: nvtxEventAttributes
+       type(nvtxEventAttributes) :: event
+     end subroutine nvtxRangePushEx
+  end interface nvtxRangePush
+
+  interface nvtxRangePop
+     subroutine nvtxRangePop() bind(C, name='nvtxRangePop')
+     end subroutine nvtxRangePop
+  end interface nvtxRangePop
+
+  ! Only the routines making up the PSyclone profiling API are public
+  public ProfileInit, ProfileFinalise, ProfileStart, ProfileEnd
+
+contains
+
+  !> An optional initialisation subroutine. This is not used for the NVTX
+  !! library.
+  subroutine ProfileInit()
+    implicit none
+    return
+  end subroutine ProfileInit
+
+  !> Starts a profiling area. The module and region name can be used to create
+  !! a unique name for each region.
+  !! Parameters: 
+  !! module_name:  Name of the module in which the region is
+  !! region_name:  Name of the region (could be name of an invoke, or
+  !!               subroutine name).
+  !! profile_data: Persistent data that can be used by the profiling library
+  !!               but isn't in this case.
+  subroutine ProfileStart(module_name, region_name, profile_data)
+    implicit none
+    character*(*), intent(in) :: module_name, region_name
+    type(ProfileData) :: profile_data
+
+    ! Round-robin the colour of each region created. Although nvtxStartRange
+    ! does a mod() on the id, we don't want to risk overflow so we don't
+    ! just blindly increment it.
+    if (last_colour < NUM_COLOURS) then
+       last_colour = last_colour + 1
+    else
+       last_colour = 1
+    end if
+    call nvtxStartRange(trim(module_name)//":"//region_name, &
```

Comment:
I usually try to avoid concatenating strings in a profile call. While I admit I have never measured it, I would suspect that this is slower than storing the concatenated name in ProfileData (together with a flag), see drhook.f90 for an example.

Reply:
Good point. I too was very careful about things like that in dl_timer but adopted a more gung-ho approach here (i.e. just copied what someone else had done). I've improved things now.

Reply:
Looks good now, thanks!